File: app/src/main/java/com/example/tradeup/data/source/remote/FirebaseTransactionSource.java
// File: src/main/java/com/example/tradeup/data/source/remote/FirebaseTransactionSource.java
package com.example.tradeup.data.source.remote;

import androidx.annotation.NonNull;
import com.example.tradeup.data.model.Transaction;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class FirebaseTransactionSource {

    private final CollectionReference transactionsCollection;

    @Inject
    public FirebaseTransactionSource(FirebaseFirestore firestore) {
        this.transactionsCollection = firestore.collection("transactions");
    }

    public Task<String> createTransaction(@NonNull Transaction transaction) {
        // Tạo một document mới với ID tự động và trả về ID đó
        return transactionsCollection.add(transaction)
                .continueWith(task -> {
                    if (!task.isSuccessful()) {
                        throw Objects.requireNonNull(task.getException());
                    }
                    return task.getResult().getId();
                });
    }

    public Task<Transaction> getTransactionById(String transactionId) {
        return transactionsCollection.document(transactionId).get()
                .continueWith(task -> {
                    if (!task.isSuccessful()) {
                        throw Objects.requireNonNull(task.getException());
                    }
                    // Firestore sẽ tự gán ID vào trường có @DocumentId
                    return task.getResult().toObject(Transaction.class);
                });
    }

    public Task<List<Transaction>> getTransactionsByUser(String userId, @NonNull String asRole, long limit) {
        return transactionsCollection
                .whereEqualTo(asRole, userId)
                .orderBy("transactionDate", Query.Direction.DESCENDING)
                .limit(limit)
                .get()
                .continueWith(task -> {
                    if (!task.isSuccessful()) {
                        throw Objects.requireNonNull(task.getException());
                    }
                    return task.getResult().toObjects(Transaction.class);
                });
    }

    public Task<Transaction> getTransactionByItemId(String itemId) {
        return transactionsCollection
                .whereEqualTo("itemId", itemId)
                .limit(1)
                .get()
                .continueWith(task -> {
                    if (!task.isSuccessful()) {
                        throw Objects.requireNonNull(task.getException());
                    }
                    QuerySnapshot querySnapshot = task.getResult();
                    if (querySnapshot != null && !querySnapshot.isEmpty()) {
                        return querySnapshot.getDocuments().get(0).toObject(Transaction.class);
                    }
                    return null;
                });
    }

    public Task<Void> updateTransaction(String transactionId, Map<String, Object> updates) {
        return transactionsCollection.document(transactionId).update(updates);
    }
}

File: app/src/main/java/com/example/tradeup/data/source/remote/FirebaseTransactionSource.kt
package com.example.tradeup.data.source.remote

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.example.tradeup.data.model.Transaction // Đảm bảo bạn có Transaction.kt
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

class FirebaseTransactionSource @Inject constructor(private val firestore: FirebaseFirestore) {

    private val transactionsCollection = firestore.collection("transactions")

    suspend fun createTransaction(transaction: Transaction): Result<String> { // Trả về transactionId
        return try {
            val documentReference = transactionsCollection.add(transaction).await()
            Result.success(documentReference.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getTransactionById(transactionId: String): Result<Transaction?> {
        return try {
            val documentSnapshot = transactionsCollection.document(transactionId).get().await()
            val transaction = documentSnapshot.toObject(Transaction::class.java)
            Result.success(transaction)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getTransactionsByUser(userId: String, asRole: String, limit: Long = 10): Result<List<Transaction>> {
        // asRole có thể là "buyerId" hoặc "sellerId" (tên field trong Transaction model)
        if (asRole !in listOf("buyerId", "sellerId")) {
            return Result.failure(IllegalArgumentException("Invalid role specified. Must be 'buyerId' or 'sellerId'."))
        }
        return try {
            val querySnapshot = transactionsCollection
                .whereEqualTo(asRole, userId)
                .orderBy("transactionDate", Query.Direction.DESCENDING)
                .limit(limit)
                .get()
                .await()
            val transactions = querySnapshot.documents.mapNotNull { it.toObject(Transaction::class.java) }
            Result.success(transactions)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}